File: src/main/java/kookmin/software/capstone2023/timebank/presentation/api/v1/BankAccountController.java
package kookmin.software.capstone2023.timebank.presentation.api.v1;

import jakarta.servlet.http.HttpServletRequest;
import kookmin.software.capstone2023.timebank.application.service.bank.account.BankAccountCreateService;
import kookmin.software.capstone2023.timebank.application.service.bank.account.BankAccountReadService;
import kookmin.software.capstone2023.timebank.application.service.bank.account.BankAccountUpdateService;
import kookmin.software.capstone2023.timebank.presentation.api.RequestAttributes;
import kookmin.software.capstone2023.timebank.presentation.api.auth.model.UserAuthentication;
import kookmin.software.capstone2023.timebank.presentation.api.auth.model.UserContext;
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.bank.account.*;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.stream.Collectors;

@UserAuthentication
@RestController
@RequestMapping("api/v1/bank/account")
public class BankAccountController {

    private final BankAccountCreateService bankAccountCreateService;
    private final BankAccountReadService bankAccountReadService;
    private final BankAccountUpdateService bankAccountUpdateService;

    public BankAccountController(BankAccountCreateService bankAccountCreateService,
                                 BankAccountReadService bankAccountReadService,
                                 BankAccountUpdateService bankAccountUpdateService) {
        this.bankAccountCreateService = bankAccountCreateService;
        this.bankAccountReadService = bankAccountReadService;
        this.bankAccountUpdateService = bankAccountUpdateService;
    }

    @PostMapping
    public BankAccountCreateResponseData createBankAccount(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext,
            @Validated @RequestBody BankAccountCreateRequestData data) {
        BankAccountCreateService.CreatedBankAccount createdBankAccount =
                bankAccountCreateService.createBankAccount(
                        userContext.getAccountId(),
                        data.getPassword(),
                        Long.valueOf(1));

        return new BankAccountCreateResponseData(
                createdBankAccount.getBalance(),
                createdBankAccount.getBankAccountNumber(),
                createdBankAccount.getBankAccountId()
        );
    }

    @GetMapping
    public List<BankAccountReadResponseData> readBankAccounts(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext) {
        List<BankAccountReadService.ReadedBankAccount> bankAccountReadResponseDataList =
                bankAccountReadService.readBankAccountsByAccountId(userContext.getAccountId());

        return bankAccountReadResponseDataList.stream()
                .map(it -> new BankAccountReadResponseData(
                        it.getBankAccountId(),
                        it.getBranchId(),
                        it.getBalance(),
                        it.getCreatedAt(),
                        it.getBankAccountNumber(),
                        it.getOwnerName(),
                        it.getOwnerType()))
                .collect(Collectors.toList());
    }

    @GetMapping("{bankAccountNumber}")
    public BankAccountReadResponseData readBankAccount(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext,
            @PathVariable String bankAccountNumber) {
        BankAccountReadService.ReadedBankAccount bankAccountReadResponseData =
                bankAccountReadService.readBankAccountByAccountNumber(
                        userContext.getAccountId(),
                        bankAccountNumber);

        return new BankAccountReadResponseData(
                bankAccountReadResponseData.getBankAccountId(),
                bankAccountReadResponseData.getBranchId(),
                bankAccountReadResponseData.getBalance(),
                bankAccountReadResponseData.getCreatedAt(),
                bankAccountReadResponseData.getBankAccountNumber(),
                bankAccountReadResponseData.getOwnerName(),
                bankAccountReadResponseData.getOwnerType());
    }

    @PostMapping("/account-password-verification")
    public PasswordVerificationResponseData verifyAccountPassword(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext,
            @Validated @RequestBody PasswordVerificationRequestData data,
            HttpServletRequest httpRequest) {
        String ipAddress = getClientIpAddress(httpRequest);
        BankAccountReadService.VerificationResult res = bankAccountReadService.verifyPassword(data, ipAddress);
        return new PasswordVerificationResponseData(
                String.valueOf(res.getFailedAttempts()),
                String.valueOf(res.getIsPasswordCorrect()));
    }

    @PutMapping("/password")
    public PasswordUpdateResponseData updatePassword(
            @RequestAttribute(RequestAttributes.USER_CONTEXT) UserContext userContext,
            @Validated @RequestBody PasswordUpdateRequestData data) {
        BankAccountUpdateService.UpdatedBankAccount updatedBankAccount =
                bankAccountUpdateService.updateBankAccountPassword(
                        userContext.getAccountId(),
                        data.getBankAccountNumber(),
                        data.getBeforePassword(),
                        data.getAfterPassword());

        return new PasswordUpdateResponseData(
                updatedBankAccount.getBankAccountNumber(),
                updatedBankAccount.getUpdatedAt(),
                "1");
    }

    private String getClientIpAddress(HttpServletRequest request) {
        String ipAddress = request.getHeader("X-FORWARDED-FOR");
        return (ipAddress == null || ipAddress.isEmpty()) ? request.getRemoteAddr() : ipAddress;
    }
}

File: src/main/java/kookmin/software/capstone2023/timebank/presentation/api/v1/BankAccountController.kt
package kookmin.software.capstone2023.timebank.presentation.api.v1

import jakarta.servlet.http.HttpServletRequest
import kookmin.software.capstone2023.timebank.application.service.bank.account.BankAccountCreateService
import kookmin.software.capstone2023.timebank.application.service.bank.account.BankAccountReadService
import kookmin.software.capstone2023.timebank.application.service.bank.account.BankAccountUpdateService
import kookmin.software.capstone2023.timebank.presentation.api.RequestAttributes
import kookmin.software.capstone2023.timebank.presentation.api.auth.model.UserAuthentication
import kookmin.software.capstone2023.timebank.presentation.api.auth.model.UserContext
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.bank.account.BankAccountCreateRequestData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.bank.account.BankAccountCreateResponseData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.bank.account.BankAccountReadResponseData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.bank.account.PasswordUpdateRequestData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.bank.account.PasswordUpdateResponseData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.bank.account.PasswordVerificationRequestData
import kookmin.software.capstone2023.timebank.presentation.api.v1.model.bank.account.PasswordVerificationResponseData
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestAttribute
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@UserAuthentication
@RestController
@RequestMapping("api/v1/bank/account")
class BankAccountController(
    private val bankAccountCreateService: BankAccountCreateService,
    private val bankAccountReadService: BankAccountReadService,
    private val bankAccountUpdateService: BankAccountUpdateService,
) {

    /***
     * @Endpoint: POST /api/v1/bank/account
     * @Description: 은행 계좌 생성 API
     * @RequestBody: BankAccountCreateRequestData
     */
    @PostMapping
    fun createBankAccount(
        @RequestAttribute(RequestAttributes.USER_CONTEXT) userContext: UserContext,
        @Validated @RequestBody
        data: BankAccountCreateRequestData,

    ): BankAccountCreateResponseData {
        val createdBankAccount: BankAccountCreateService.CreatedBankAccount =
            bankAccountCreateService.createBankAccount(
                accountId = userContext.accountId,
                password = data.password,
                branchId = 1,
            )

        return BankAccountCreateResponseData(
            balance = createdBankAccount.balance,
            accountNumber = createdBankAccount.bankAccountNumber, // 계좌번호
            bankAccountId = createdBankAccount.bankAccountId, // 은행 계좌 id
        )
    }

    /***
     * @Endpoint: GET /api/v1/bank/account
     * @Description: 은행 계좌 목록 조회 API
     * @RequestBody: BankAccountReadRequestData
     */
    @GetMapping
    fun readBankAccounts(
        @RequestAttribute(RequestAttributes.USER_CONTEXT) userContext: UserContext,
    ): List<BankAccountReadResponseData> {
        val bankAccountReadResponseDataList: List<BankAccountReadService.ReadedBankAccount> =
            bankAccountReadService.readBankAccountsByAccountId(
                accountId = userContext.accountId,
            )

        return bankAccountReadResponseDataList.map {
            BankAccountReadResponseData(
                bankAccountId = it.bankAccountId,
                branchId = it.branchId,
                balance = it.balance,
                createdAt = it.createdAt,
                bankAccountNumber = it.bankAccountNumber,
                ownerName = it.ownerName,
                ownerType = it.ownerType,
            )
        }
    }

    /***
     * @Endpoint: GET /api/v1/bank/account/{bankAccountNumber}
     * @Description: 은행 계좌 조회 API
     * @RequestParam: bankAccountNumber
     */
    @GetMapping("{bankAccountNumber}")
    fun readBankAccount(
        @RequestAttribute(RequestAttributes.USER_CONTEXT) userContext: UserContext,
        @PathVariable bankAccountNumber: String,
    ): BankAccountReadResponseData {
        val bankAccountReadResponseData: BankAccountReadService.ReadedBankAccount =
            bankAccountReadService.readBankAccountByAccountNumber(
                accountId = userContext.accountId,
                bankAccountNumber = bankAccountNumber,
            )
        return BankAccountReadResponseData(
            bankAccountId = bankAccountReadResponseData.bankAccountId,
            branchId = bankAccountReadResponseData.branchId,
            balance = bankAccountReadResponseData.balance,
            createdAt = bankAccountReadResponseData.createdAt,
            bankAccountNumber = bankAccountReadResponseData.bankAccountNumber,
            ownerName = bankAccountReadResponseData.ownerName,
            ownerType = bankAccountReadResponseData.ownerType,
        )
    }

    /***
     * @Endpoint: POST /api/v1/bank/account-password-verification
     * @Description: 은행 계좌 비밀 번호 검증 API
     */
    @PostMapping("/account-password-verification")
    fun verifyAccountPassword(
        @RequestAttribute(RequestAttributes.USER_CONTEXT) userContext: UserContext,
        @Validated @RequestBody
        data: PasswordVerificationRequestData,
        httpRequest: HttpServletRequest,
    ): PasswordVerificationResponseData {
        val ipAddress = getClientIpAddress(httpRequest)
        val res = bankAccountReadService.verifyPassword(data, ipAddress)
        return PasswordVerificationResponseData(
            resResultDesc = res.failedAttempts.toString(),
            resResultCode = res.isPasswordCorrect.toString(),
        )
    }

    /***
     * @Endpoint: PUT /api/v1/bank/account/password
     * @Description: 은행 계좌 비밀 번호 변경 API
     */
    @PutMapping("/password")
    fun updatePassword(
        @RequestAttribute(RequestAttributes.USER_CONTEXT) userContext: UserContext,
        @Validated @RequestBody
        data: PasswordUpdateRequestData,
    ): PasswordUpdateResponseData {
        val updatedBankAccount = bankAccountUpdateService.updateBankAccountPassword(
            accountId = userContext.accountId,
            accountNumber = data.bankAccountNumber,
            beforePassword = data.beforePassword,
            afterPassword = data.afterPassword,
        )

        return PasswordUpdateResponseData(
            bankAccountNumber = updatedBankAccount.bankAccountNumber,
            updateAt = updatedBankAccount.updatedAt,
            resResultCode = "1",
        )
    }

    private fun getClientIpAddress(request: HttpServletRequest): String {
        val ipAddress = request.getHeader("X-FORWARDED-FOR")
        return if (ipAddress.isNullOrEmpty()) {
            request.remoteAddr
        } else {
            ipAddress
        }
    }
}
